### ISSUE TITLE
Generic_type_queries

### ISSUE THREAD

**Issue by Ervinoreo** [nus-cs2030/2223-s2#79]
# Summary

I have a question regarding the binding of E.

## Description:

In the lecture when we have:
static < E > ImList< E > of() {return new ImList< E >();} 
in the class ImList< E > {...} and we said that the < E > in the factory method above refers to ImList< Integer >of()'s Integer instead of the class level < E >. So in my understanding the < E > in the static method is different from the < E > of class ImList< E >'s < E >. Then my question arises, how do we parameterized the class level < E >?

## Screenshots

Screenshots (if any)
![WhatsApp Image 2023-02-06 at 20 34 04](https://user-images.githubusercontent.com/110983457/216972653-e7c627ec-ea59-4c62-8fce-a080ed7a181b.jpeg)

**Comment by Ervinoreo** [comment]
sorry it should be labelled as lecture 05

**Comment by 0liverkong** [comment]
for example if the Imlist() constructor was public instead then you could do Imlist<Integer> testlist = new Imlist<Integer>(); this would be parameterising the class level <E> as this <E> is dependent on each instance of Imlist, and hence is not the same <E> referred to by static method of().

**Comment by Ervinoreo** [comment]
> for example if the Imlist() constructor was public instead then you could do Imlist<Integer> testlist = new Imlist<Integer>(); this would be parameterising the class level <E> as this <E> is dependent on each instance of Imlist, and hence is not the same <E> referred to by static method of().

What if in the case of the above screenshot scenario when it is private, how can i parameterise the class level E, because other methods such as add(), set(), their E are all dependent on the class level E right? Is it like the new ImList< E > in the static method got the E from method level and then pass it into the class level?

**Comment by 0liverkong** [comment]
witnessing of the type in that case happens in when you call ImList.\<Integer\>.of() in your own code. by changing the type in the angle brackets within your own you control the variable types that Imlist.of() returns.

add(), remove(), get() are functions that may use the generics in the class definition, but based on your implementation of ImList you should have more than enough info of the type of variables in that your ImList instance can store.This willl allow you to do stuff like storing the get() return value in a variable, whose type you should know.

**Comment by Ervinoreo** [comment]
First of all, thank you for your answer!

so it is like when we make ImList.< Integer >.of(), it will pass into method level E, but then it will return me an instance of say ImList< Integer >, so when we are trying to use non static method via this object, the E of the non static method will bind to this < Integer > that is being returned. 

It actually makes me think of the new ImList < E > that is being returned by the static method, will first accept the E from method level and pass into class level. I wonder could I think it this way tho?

**Comment by 0liverkong** [comment]
yup thanks for clearing up my haphazard phrasing, that was what i meant :D.

for the second part i Think thats okay.....? not sure if thats how it actually works.
